### ISSUE TITLE
Can't connect (SCP, SSH) in Xamarin Android

### ISSUE THREAD

**Issue by stealthrabbi** [sshnet/SSH.NET#807]
Calling `Connect()` with an SSH or SCP client (and likely others) causes this exception. Does this library not work in Xamarin Android?

I don't see where in this library ti's throwing `NotImplementedException`, so I'm guessing something missing in Mono / .NET Standard?

```
2021-04-15 12:29:05.030 16384-16427/? E/MySessionStarter: Failed uploading file: System.NotImplementedException: The method or operation is not implemented.
      at Renci.SshNet.Session.WaitOnHandle (System.Threading.WaitHandle waitHandle, System.TimeSpan timeout) [0x00041] in <c7169cd606324a41b931acc40c516428>:0 
      at Renci.SshNet.Session.WaitOnHandle (System.Threading.WaitHandle waitHandle) [0x0000d] in <c7169cd606324a41b931acc40c516428>:0 
      at Renci.SshNet.Session.Connect () [0x0017b] in <c7169cd606324a41b931acc40c516428>:0 
      at Renci.SshNet.BaseClient.CreateAndConnectSession () [0x00053] in <c7169cd606324a41b931acc40c516428>:0 
      at Renci.SshNet.BaseClient.Connect () [0x0001f] in <c7169cd606324a41b931acc40c516428>:0 
    ....
    ```

**Comment by drwtsn32x** [comment]
Did you solve this yet? What version of mono are you using?

**Comment by stealthrabbi** [comment]
Not solved. I don't know what version of Mono I'm using. I believe it was roughly the latest Xamarin.

**Comment by drwtsn32x** [comment]
When you get a chance, if you could check the mono version that'd be great....

```
$ mono --version
```

I'm exploring possibly the same issue with a different product that uses the SSH.NET library. Wondering if the mono version is the culprit.

**Comment by stealthrabbi** [comment]
What you say about the keys may be right, but I think my issue is happening way before a connection. It's related to `Session.WaitOnHandle` being not implemented.

**Comment by ts678** [comment]
> It's related to Session.WaitOnHandle being not implemented.

It seems Session.WaitOnHandle is in [https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Session.cs](https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Session.cs)
You can see comments there about the receive thread. I'm not an SSH.NET developer. It'd be nice if they can help.

It might be waiting at the below for the receive thread to either receive, or find an exception, and it got exception:

https://github.com/sshnet/SSH.NET/blob/f072c5f1a1bb607797a85cd43cf498ec541af3ab/src/Renci.SshNet/Session.cs#L852-L852

> I think my issue is happening way before a connection

The stack trace names suggest that it's working on connect, but it likely died before SSH connection was finished.
The behavior we see with certain SFTP servers such as Bitvise (but not OpenSSH, at least so far) is a failure after a

Server: Elliptic Curve Diffie-Hellman Key Exchange Reply, New Keys
Message Code: Elliptic Curve Diffie-Hellman Key Exchange Reply (31)
is received, containing
KEX host key (type: ecdsa-sha2-nistp384)
I think the received protocol is RFC 5656 ECDH Key Exchange [here](https://datatracker.ietf.org/doc/html/rfc5656#section-4).

Processing of that server `SSH_MSG_KEX_ECDH_REPLY` might be here:

https://github.com/sshnet/SSH.NET/blob/f072c5f1a1bb607797a85cd43cf498ec541af3ab/src/Renci.SshNet/Session.cs#L1298-L1302

and at some point the key from the server might hit here:

https://github.com/sshnet/SSH.NET/blob/a5bd08d655bb6a3c762306472cec354556dca3a3/src/Renci.SshNet/Security/Cryptography/EcdsaKey.cs#L238-L242

[ECDsa Class](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.ecdsa) and [ECDsaCng Class](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.ecdsacng) get involved, and so maybe it hits the NotImplementedException around [here](https://github.com/mono/mono/blob/def4f3be68c44c3fb353fe46f00d540f4ed9bd15/mcs/class/referencesource/System.Core/System/Security/Cryptography/ECDsaCng.cs#L15-L36).

NotImplementedException has been in ECDsaCng.cs a long time, maybe added [here](https://github.com/mono/mono/commit/3c3ef17376f5b4e71ac024210672e2d0484f3419), then later exposed through
[Support newer SSH Ciphers and MACs #53](https://github.com/sshnet/SSH.NET/issues/53#issuecomment-621711274). If you're still on the 2016 release, ignore all this. It applies to 2020's.

Any SSH.NET developer comments? Could these very welcome new capabilities be exposing an old gap in mono?
Please forgive any mistakes in the details. I don't have a debugger, or know the protocol or the code all that well.

**Comment by ts678** [comment]
I'm noticing the 2017 mono change did a similar NotImplementedException to RsaCng.cs just below ECDsaCng.cs

**Comment by drwtsn32x** [comment]
> (but not OpenSSH, at least so far)

I was able to reproduce with an OpenSSH server by using an ECDSA key.

**Comment by Joannick** [comment]
Hi friend,
so did you found a solution about this? I've seen an other solution is Rebex SFTP Nugget but licence is 350$.
Thanks

**Comment by stealthrabbi** [comment]
My solution was in Xamarin, to create a binding to the JSCH java library and use that in my xamarin android proj.

**Comment by ts678** [comment]
A workaround is try new ECDsaCng() to see if it's implemented. If exception occurs, go through HostKeyAlgorithms.Keys, and HostKeyAlgorithms.Remove(key) the ones beginning with ecdsa (inspired by [Limit what algorithms or ciphers are used #730](https://github.com/sshnet/SSH.NET/issues/730)).

**Comment by TechnoFairyGirl** [comment]
ECDSA is simply [not implemented](https://github.com/mono/mono/blob/main/mcs/class/referencesource/System.Core/System/Security/Cryptography/ECDsaCng.cs) on Mono. The only real fix would be to use a different implementation, such as the one provided by Bouncy Castle.

You can work around the issue by removing the ECDSA algorithms from `connectionInfo.HostKeyAlgorithms` if `System.Security.Cryptography.ECDsaCng` is unavailable.

```csharp
try { using (var ecdsa = new System.Security.Cryptography.ECDsaCng()) ; }
catch (NotImplementedException)
{
    var algsToRemove = connectionInfo.HostKeyAlgorithms.Keys.Where(algName => algName.StartsWith("ecdsa")).ToArray();
    foreach (var algName in algsToRemove) connectionInfo.HostKeyAlgorithms.Remove(algName);
}
```

If you get `Renci.SshNet.Common.SshConnectionException: 'An established connection was aborted by the server.'` on connect after adding the code above, then the server doesn't have any mutually agreeable host key algorithms now that you've removed the ECDSA ones. This can be solved on the server by adding `ssh-rsa` to `HostKeyAlgorithms` in `sshd_config` (or the equivalent for whatever SSH server you have).

**Comment by Rob-Hague** [comment]
Should be fixed by #1461
